logstash-core/lib/logstash/pipelines_registry.rb

```diff
@@ -17,40 +17,99 @@
 
 module LogStash
   class PipelineState
-    attr_reader :pipeline_id, :pipeline
+    attr_reader :pipeline_id
 
     def initialize(pipeline_id, pipeline)
       @pipeline_id = pipeline_id
       @pipeline = pipeline
-      @reloading = Concurrent::AtomicBoolean.new(false)
+      @loading = Concurrent::AtomicBoolean.new(false)
+
+      # this class uses a lock to ensure thread safe visibility.
+      @lock = Mutex.new
     end
 
     def terminated?
-      # a reloading pipeline is never considered terminated
-      @reloading.false? && @pipeline.finished_execution?
+      @lock.synchronize do
+        # a loading pipeline is never considered terminated
+        @loading.false? && @pipeline.finished_execution?
+      end
     end
 
-    def set_reloading(is_reloading)
-      @reloading.value = is_reloading
+    def set_loading(is_loading)
+      @lock.synchronize do
+        @loading.value = is_loading
+      end
     end
 
     def set_pipeline(pipeline)
-      raise(ArgumentError, "invalid nil pipeline") if pipeline.nil?
-      @pipeline = pipeline
+      @lock.synchronize do
+        raise(ArgumentError, "invalid nil pipeline") if pipeline.nil?
+        @pipeline = pipeline
+      end
+    end
+
+    def pipeline
+      @lock.synchronize { @pipeline }
     end
   end
 
+  class PipelineStates
+
+    def initialize
+      @states = {}
+      @locks = {}
+      @lock = Mutex.new
+    end
+
+    def get(pipeline_id)
+      @lock.synchronize do
+        @states[pipeline_id]
+      end
+    end
+
+    def put(pipeline_id, state)
+      @lock.synchronize do
+        @states[pipeline_id] = state
+      end
+    end
+
+    def remove(pipeline_id)
+      @lock.synchronize do
+        @states.delete(pipeline_id)
+      end
+    end
+
+    def size
+      @lock.synchronize do
+        @states.size
+      end
+    end
+
+    def empty?
+      @lock.synchronize do
+        @states.empty?
+      end
+    end
+
+    def each_with_object(init, &block)
+      states = @lock.synchronize { @states.dup }
+      states.each_with_object(init, &block)
+    end
+
+    def get_lock(pipeline_id)
+      @lock.synchronize do
+        @locks[pipeline_id] ||= Mutex.new
+      end
```

Comment:
Optional: avoid acquiring a lock when the value is already populated

```suggestion
      @locks.fetch(pipeline_id) do |h,k|
        @lock.synchronize do
          h[k] ||= Mutex.new
        end
      end
```

***

Nitpick: we seem to have two concerns that each require their own mutual exclusion, and could benefit from having separate locks for state-operations and lock-acquisition.

Reply:
In general the very first thing you have to do is to acquire the global lock because when `synchronize` is not yet called by this thread, the *happens-before* relationship is not established and what this thread sees in the `@locks` collection might not reflect what other thread have written to it.

Your suggestion should actually be 

```ruby
      @locks.fetch(pipeline_id) do |k|
        @lock.synchronize do
          @locks[k] ||= Mutex.new
        end
      end
```

But this will only work because you actually do the `||= Mutex.new` within the `synchronize` block. As I mentioned above, the actual state of the `@locks` collection is not yet known until we synchronize. It will also only work if object are strictly added to the collection. If we remove objects from @locks, that code could return a removed state in the `fetch` call. 

IMO this is an unnecessary optimization. This is nowhere near a hot path that could possibly from this optimization and it will not be thread safe if we eventually add objects removal from the `@locks` collection.